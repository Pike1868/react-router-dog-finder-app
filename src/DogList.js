import React from "react";
import { Link } from "react-router-dom";

/**
 * DogList Component:
 *
 * Returns each dog displayed in a div, with its image, and name as a Link to that dogs detail page
 *
 * Props:
 * - dogs (array): Array of dog objects
 *
 */

const DogList = ({ dogs }) => {
  return (
    <div className="DogList">
      <div className="row mt-4">
        <div className="col">
          <h2 className="text-center">
            HELLO. WE HAVE DOGS. CLICK ON THEM FOR MORE INFO.
          </h2>
        </div>
      </div>
      <div className="row">
        {dogs.map((d) => (
          <div className="col-3 text-center" key={d.name}>
            <img src={`/${d.src}.jpg`} alt={d.name} />
            <h3 className="mt-3">
              <Link to={`/dogs/${d.name.toLowerCase()}`}>{d.name}</Link>
            </h3>
          </div>
        ))}
      </div>
    </div>
  );
};

export default DogList;
