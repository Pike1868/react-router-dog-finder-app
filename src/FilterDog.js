import React, { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import DogDetails from "./DogDetails";

/**
 * FilterDog Component:
 * 
 * Responsible for determining if the requested dog exists based on URL parameter
 * If dog exists, renders the DogDetails component with the found dog's data.
 * If dog does NOT exist, uses a timer to display an error message, then redirects back to the DogList component
 * 
 * Props:
 * - dogs (array): Array of dog objects
 *
 * State:
 * - redirect (boolean): Controls the display of the redirect message and the initiation of redirection.
 */

const FilterDog = ({ dogs }) => {
  const navigate = useNavigate();
  const { name } = useParams();
  const [redirect, setRedirect] = useState(false);

  const existingDog = dogs.find(
    (d) => d.name.toLowerCase() === name.toLowerCase()
  );

  useEffect(() => {
    if (!existingDog) {
      const timer = setTimeout(() => {
        navigate("/dogs");
      }, 3000);
      setRedirect(true);
      return () => clearTimeout(timer);
    }
  }, [name, existingDog, navigate]);

  if (redirect) {
    return (
      <p>
        No dog found with the name of <strong>{name}</strong>. Redirecting to
        all dogs....
      </p>
    );
  }

  if (existingDog) {
    return <DogDetails dog={existingDog} />;
  }
};

export default FilterDog;
