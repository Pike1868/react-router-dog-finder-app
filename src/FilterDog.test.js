import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { useParams, useNavigate } from "react-router-dom";
import FilterDog from "./FilterDog";
import DogList from "./DogList";
import dogs from "./dogs";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
  useNavigate: jest.fn(),
}));

//Smoke Test
test("FilterDog component renders without crashing for an existing dog", () => {
  useParams.mockReturnValue({ name: "whiskey" });
  useNavigate.mockReturnValue(jest.fn());

  render(
    <MemoryRouter initialEntries={[`/dogs/whiskey`]}>
      <FilterDog dogs={dogs} />
    </MemoryRouter>
  );
});

//Snapshot Test
test("FilterDog component matches snapshot", () => {
  useParams.mockReturnValue({ name: "whiskey" });
  useNavigate.mockReturnValue(jest.fn());

  const { asFragment } = render(
    <MemoryRouter initialEntries={[`/dogs/whiskey`]}>
      <FilterDog dogs={dogs} />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

// Test for existing dog
test("renders DogDetails for an existing dog", () => {
  useParams.mockReturnValue({ name: "whiskey" });
  useNavigate.mockReturnValue(jest.fn());

  render(
    <MemoryRouter initialEntries={[`/dogs/whiskey}`]}>
      <Routes>
        <Route path="/dogs/:name" element={<FilterDog dogs={dogs} />} />
      </Routes>
    </MemoryRouter>
  );
  expect(screen.getByText(`Meet Whiskey!`)).toBeInTheDocument();
});

// Test for non-existing dog
test("displays error message for a non-existing dog", () => {
  useParams.mockReturnValue({ name: "random" });
  useNavigate.mockReturnValue(jest.fn());

  render(
    <MemoryRouter initialEntries={[`/dogs/random`]}>
      <Routes>
        <Route path="/dogs/:name" element={<FilterDog dogs={dogs} />} />
        <Route path="/dogs" element={<DogList dogs={dogs} />} />
      </Routes>
    </MemoryRouter>
  );

  expect(
    screen.getByText(
      `No dog found with the name of random. Redirecting to all dogs....`
    )
  ).toBeInTheDocument();
});
