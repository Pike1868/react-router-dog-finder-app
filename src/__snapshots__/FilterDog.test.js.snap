// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDog component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Meet Whiskey!
    </h2>
    <div
      class="col-3 text-center"
    >
      <img
        alt="Whiskey"
        src="/whiskey.jpg"
      />
    </div>
    <ul>
      <li>
        Whiskey loves eating popcorn.
      </li>
      <li>
        Whiskey is a terrible guard dog.
      </li>
      <li>
        Whiskey wants to cuddle with you!
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
